src/api.v2/helpers/worker/workSubmit/createWorkerK8s.js

```diff
@@ -62,7 +104,14 @@ const createWorkerResources = async (service) => {
     return { name, creationTimestamp, phase };
   }
 
-  const pods = await getAvailablePods(namespace);
+  // scale pods if worker has zero replicas
+  let pods = await getAvailablePods(namespace);
+  const { spec: { replicas } } = await getDeployment('worker', namespace);
+  if (pods.length < 1 && replicas === 0) {
```

Comment:
In this if statement: `if (pods.length < 1 && replicas === 0)` about this part: `pods.length < 1`. Is 1 the value of desired replicas? If so, I suggest creating a new variable called `desiredReplicas` that you can also pass in line 111 when you are calling function `scaleDeploymentReplicas`, so that we avoid hardcoding 1.

Reply:
About this 

> I originally had the check in scaleDeploymentReplicas for the reason you mentioned but felt that scaleDeploymentReplicas should be explicitly for scaling the replicas as opposed to also checking if the condition is such that we want to scale. The idea being the it is slightly confusing to always call scaleDeploymentReplicas when in many cases it will not scale the replicas. Happy to change though if you think this is outweighed by the added simplicity or is not a concern.

I understand what you mean. Could we get the deployment once, as you are already getting it on line 108 (`const { spec: { replicas } } = await getDeployment('worker', namespace);`), and then instead of getting the deployment a second time in `scaleDeploymentReplicas`, pass it as a parameter to `scaleDeploymentReplicas`? My thinking here is that it avoids the possible confusion that you outlined, which is a good point, and saves 1 extra call to getDeployment.

Reply:
Honestly it just makes things a mess when I try to do this and adds more lines than it eliminates. If you like this better I will go for it but it looks super ugly to me. If you have an alternative happy to do that also:

**deployment as parameter**:

```javascript
const scaleDeploymentReplicas = async (name, namespace, deployment, desiredReplicas = 1) => {
  const k8sApi = kc.makeApiClient(k8s.AppsV1Api);

  logger.log(`Scaling ${name} from ${deployment.spec.replicas} to ${desiredReplicas} replicas...`);

  // edit
  // eslint-disable-next-line no-param-reassign
  deployment.spec.replicas = desiredReplicas;

  // replace
  await k8sApi.replaceNamespacedDeployment(name, namespace, deployment);

  // wait until we have pods
  await waitForPods(namespace);
};


....
// scale pods if worker has zero replicas
  let replicas;
  let deployment;
  const minDesiredReplicas = 1;
  let pods = await getAvailablePods(namespace);

  try {
    deployment = await getDeployment('worker', namespace);
    replicas = deployment.spec.replicas;
  } catch (e) {
    logger.log('Getting worker deployment failed, ignoring...', e);
  }

  if (pods.length < 1 && replicas < minDesiredReplicas) {
    await scaleDeploymentReplicas('worker', namespace, deployment, minDesiredReplicas);
    pods = await getAvailablePods(namespace);
  }
  ...
```

**call getDeployment twice**:

```javascript

const scaleDeploymentReplicas = async (name, namespace, desiredReplicas = 1) => {
  const k8sApi = kc.makeApiClient(k8s.AppsV1Api);

  const deployment = await getDeployment(name, namespace);
  logger.log(`Scaling ${name} from ${deployment.spec.replicas} to ${desiredReplicas} replicas...`);

  // edit
  deployment.spec.replicas = desiredReplicas;

  // replace
  await k8sApi.replaceNamespacedDeployment(name, namespace, deployment);

  // wait until we have pods
  await waitForPods(namespace);
};

...
// scale pods if worker has zero replicas
  let replicas;
  const minDesiredReplicas = 1;
  let pods = await getAvailablePods(namespace);
  try {
    ({ spec: { replicas } } = await getDeployment('worker', namespace));
  } catch (e) {
    logger.log('Getting worker deployment failed, ignoring...', e);
  }

  if (pods.length < 1 && replicas < minDesiredReplicas) {
    await scaleDeploymentReplicas('worker', namespace, minDesiredReplicas);
    pods = await getAvailablePods(namespace);
  }
  ...
```

Reply:
My understanding is that the code should rather look like this:

```
const scaleDeploymentReplicas = async (name, namespace, deployment, desiredReplicas = 1) => {
  const k8sApi = kc.makeApiClient(k8s.AppsV1Api);

  logger.log(`Scaling ${name} from ${deployment.spec.replicas} to ${desiredReplicas} replicas...`);

  // edit
  // eslint-disable-next-line no-param-reassign
  deployment.spec.replicas = desiredReplicas;

  // replace
  await k8sApi.replaceNamespacedDeployment(name, namespace, deployment);

  // wait until we have pods
  await waitForPods(namespace);
};

....
// scale pods if worker has zero replicas
  const minDesiredReplicas = 1;
  let pods = await getAvailablePods(namespace);

  try {
    const deployment = await getDeployment('worker', namespace);
    const replicas = deployment.spec.replicas;
    if (pods.length < 1 && replicas < minDesiredReplicas) {
      await scaleDeploymentReplicas('worker', namespace, deployment, minDesiredReplicas);
    }
  } catch (e) {
    logger.log('Getting worker deployment failed, ignoring...', e);
  }

   pods = await getAvailablePods(namespace);
  ...
```

I moved `pods = await getAvailablePods(namespace);` outside of the if statement, because it currently is executed like this, from what I can see in the changes diff. `if (pods.length < 1 && replicas < minDesiredReplicas)` should be in the try and catch, because it is coupled with getting `deployment` -- if `getDeployment` errors out or returns nothing, then even if the if statement is true, it will not execute with correct value of deployment. Am I missing something? What do you think about this code?

Reply:
It won't error because `replicas < minDesiredReplicas` will still evaluate to false as replicas has been instantiated (`let replicas;`). I like the above as it gets rid of some of this instantiation though

Reply:
For the `getAvailablePods` it only runs it once currently. The second call is directly tied to the scaling as it needs to be called again after scaling --> I think it makes sense to keep it inside the if statement

Reply:
my mistake on the `getAvailablePods` -- you are right

Reply:
thanks for updating the code!